src/bin/nydus-image/deduplicate.rs

```diff
@@ -0,0 +1,661 @@
+// Copyright (C) 2023 Nydus Developers. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Deduplicate for Chunk.
+use anyhow::{Context, Result};
+use nydus_api::ConfigV2;
+use nydus_builder::Tree;
+use nydus_rafs::metadata::RafsSuper;
+use nydus_storage::device::BlobInfo;
+use rusqlite::{params, Connection};
+use std::fs;
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+
+#[derive(Debug)]
+pub enum DatabaseError {
+    SqliteError(rusqlite::Error),
+    PoisonError(String),
+    // Add other database error variants here as needed, e.g.:
+    // MysqlError(mysql::Error),
+}
+
+impl std::fmt::Display for DatabaseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            DatabaseError::SqliteError(ref err) => err.fmt(f),
+            DatabaseError::PoisonError(ref err) => write!(f, "PoisonError: {}", err),
+            // Add other error type formatting here
+        }
+    }
+}
+
+impl std::error::Error for DatabaseError {}
+
+impl From<rusqlite::Error> for DatabaseError {
+    fn from(error: rusqlite::Error) -> Self {
+        DatabaseError::SqliteError(error)
+    }
+}
+
+pub trait Database {
+    /// Creates a new chunk in the database.
+    fn create_chunk_table(&self) -> Result<()>;
+
+    /// Creates a new blob in the database.
+    fn create_blob_table(&self) -> Result<()>;
+
+    /// Inserts chunk information into the database.
+    fn insert_chunk(&self, chunk_info: &Chunk) -> Result<()>;
+
+    /// Inserts blob information into the database.
+    fn insert_blob(&self, blob_info: &Blob) -> Result<()>;
+
+    /// Retrieves all chunk information from the database.
+    fn get_chunks(&self) -> Result<Vec<Chunk>>;
+
+    /// Retrieves all blob information from the database.
+    fn get_blobs(&self) -> Result<Vec<Blob>>;
+}
+
+pub struct SqliteDatabase {
+    chunk_table: Option<ChunkTable>,
```

Comment:
Why do we need to wrap into `Option` here?

Reply:
Historical legacy issue resolved.